### DISCUSSION TITLE
Directory not Empty Error

### DISCUSSION THREAD

**Discussion by funkydan2** [Homebrew/discussions#633] [Everyday usage]
Following Perl being upgraded from 5.32.0 to 5.32.1 I got the following error:
```
Removing: /usr/local/Cellar/perl/5.32.0... (3,903 files, 92.9MB)
Error: Directory not empty @ dir_s_rmdir - /usr/local/Cellar/perl/5.32.0
```
There are many files and directories in the `5.32.0` directory. Is it safe to delete it using `rm -r` or is there a `brew` command I should try first?

**Comment by carlocab** [comment]
Try to uninstall `perl` fully first (you way need to do `brew uninstall` more than once) and then install it again first. There's a chance that the links in `/usr/local/bin`, et al still point to your `5.32.0` installation, so just removing the directory will make Homebrew `perl` in accessible for you.

**Reply by funkydan2** [reply]
Thanks @carlocab.

Uninstalling lead to this error:
```
Uninstalling perl... (2,230 files, 47.9MB)
Error: Could not remove perl keg! Do so manually:
  sudo rm -rf /usr/local/Cellar/perl/5.32.0
```
so it seems that `rm -rf` is the way forward anyway!

**Reply by carlocab** [reply]
Hmm. I see. Do `brew unlink perl` before removing that, and then, after deleting this, make sure you don't have any remaining Homebrew `perl` installs lying around before reinstalling.

**Comment by emjayess** [comment]
Bumped up against this twice (nginx, dnsmasq) today after upgrading package versions, and it seems like a pretty obtuse error...

```zsh
Removing: /opt/homebrew/Cellar/dnsmasq/2.84... (5 files, 498.7KB)
Error: Directory not empty @ dir_s_rmdir - /opt/homebrew/Cellar/dnsmasq/2.84
```

... because the directory not being empty is _why_ it ought to be cleaned up; like, do you wash your dishes before putting them in the dishwasher? 🙄

**Reply by frknasir** [reply]
how did you solve it? like above?

**Reply by roncodes** [reply]
I wash my dishes before putting them in the dishwasher... double clean 😅

**Comment by Truzze** [comment]
This error happened to me about 5 times now with different packages (including: php, dnsmasq, python, nginx, ...)

For me this error happens because the `bin` and/or `sbin` directory inside those package directories is owned by `root` instead of my user like every other directory inside the package directories.

To fix the issue just **go into the reported directory** and change ownership of the `bin` and `sbin` directory to your user.
`sudo chown -R "$USER":admin bin`
`sudo chown -R "$USER":admin sbin` sometimes the `sbin` directory does not exists so you can skip this one 

Now your previous command you where trying to run should work without the error

**Reply by SMillerDev** [reply]
> For me this error happens because the bin and/or sbin directory inside those package directories is owned by root instead of my user like every other directory inside the package directories.

This can be caused by using brew services with those (Which will display a warning about it when you use it with sudo)

**Reply by ijoseph** [reply]
In general, you can use [`fd`](https://github.com/sharkdp/fd) to find files owned by `root` with `-o root`, e.g.:

```bash
fd  -o root   /usr/local/Cellar/ 
```

probably also with `find` too, but I've grown accustomed to `fd` so don't know the syntax off the top of my head for `find`.

In my case, the offending files were `__pycache__` `.pyc` files associated with `glances`, probably because I ran `sudo glances` at some point which meant that `root` compiled those `.py` files.

**Reply by gromgit** [reply]
>```
> fd  .  -o root   /usr/local/Cellar/ 
>```

Not sure what that period does, but the `find` equivalent is:
```
find /usr/local/Cellar -user root
```

**Reply by nathaliachn** [reply]
> This error happened to me about 5 times now with different packages (including: php, dnsmasq, python, nginx, ...)
> 
> For me this error happens because the `bin` and/or `sbin` directory inside those package directories is owned by `root` instead of my user like every other directory inside the package directories.
> 
> To fix the issue just **go into the reported directory** and change ownership of the `bin` and `sbin` directory to your user.
> `sudo chown -R "$USER":admin bin`
> `sudo chown -R "$USER":admin sbin` sometimes the `sbin` directory does not exists so you can skip this one
> 
> Now your previous command you where trying to run should work without the error

It help! thanks you :)

**Reply by vfsoraki** [reply]
This fixed issue for me too. I ran this as mentioned by @gromgit, added `chown` to that.

```
find /opt/homebrew/Cellar -user root | xargs sudo chown -R $USER
```

I'm on M1, so mine was at `/opt/homebrew`. Your path might be different.

**Comment by ElForastero** [comment]
It looks like I had laravel/vallet installed. It installs nginx, php and dnsmasq - exactly the problematic packages.

**Reply by emjayess** [reply]
Hm I also have [`valet`](https://github.com/laravel/valet). Does it `brew install` these packages? And will it detect and use these packages locally, if they're already installed?

**Comment by MincePie** [comment]
Did you figure this out? I have a similar problem now with node. I lost access to npx recently (I think there may be something with version control on node v 16). I tried to remove node and npm from by usr/local/bin and when I try to brew upgrade (to add them back) I get an error that says:Error: Directory not empty @ dir_s_rmdir - /usr/local/Cellar/node/7.9.0. I don't know what this means or how to address it. Does anyone know how to resolve this?

**Reply by carlocab** [reply]
If you no longer need your install of Node 7.9, you can do 
```
sudo rm -rf /usr/local/Cellar/node/7.9.0
```

**Comment by eNipu** [comment]
For me, it usually happens while  ```brew update``` or  ```brew cleanup```.
The following permission change work out for me 
`sudo chown -R "$USER" "$(brew --prefix)/Cellar"`

**Reply by yamenarahman** [reply]
That's it, Thanks!

**Reply by tmillr** [reply]
Cellar is already owned by my user, but the issue persists.

```sh
==> `brew cleanup` has not been run in the last 30 days, running now...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Removing: ~/Library/Caches/Homebrew/pango--1.50.6... (773.8KB)
Removing: /usr/local/Cellar/sqlite/3.38.5... (11 files, 4.4MB)
Removing: ~/Library/Caches/Homebrew/sqlite--3.38.5... (2MB)
Error: Directory not empty @ dir_s_rmdir - /usr/local/lib/node_modules/@typescript-eslint/.DS_Store
```

**Reply by SMillerDev** [reply]
Is that particular file owned by your user too?

**Reply by tmillr** [reply]
> Is that particular file owned by your user too?

Hmmm I'm not 100% sure. I ended up just `rm -rf` the entire node module/pkg and then retrying and it worked then (I believe I was trying to install something, but this error kept occurring halfway through and was fatal and kept it from installing). I believe both the the DS_Store file and it's containing dir were owned by my user, but I'd have to go back through my backup to be sure. It was not the only file in the dir either (there was another dir inside eslint for estree, which also had contents from what I recall).

**Reply by yamenarahman** [reply]
For Mac silicon:
sudo chown -R "$USER" "$(brew --prefix)/"

**Comment by eshack94** [comment]
This issue is still occurring for me on Big Sur v. 11.6. It only seems to apply to several packages in my case, including `python@3.9` and `perl`. The above workarounds fix the issue, but it would be great if the root problem could be addressed in a way that doesn't require user intervention.

**Reply by carlocab** [reply]
The root problem is likely that file/directory permissions (or ownership) changed somewhere in between a `brew install` and `brew cleanup` (which is often run implicitly by a `brew install` or `brew upgrade`).

This isn't something Homebrew does, and is typically done by the user (possibly unknowingly), so I'm not sure there's much for `brew` to do here other than perhaps failing more gracefully and displaying better error messages. Those would still require user intervention.

Open to ideas, though.

**Comment by DiegoBetancourtC** [comment]
I have the same issue, but none of the above fix it:

`==> Installing colima dependency: openssl@1.1`
`==> Pouring openssl@1.1--1.1.1m.big_sur.bottle.tar.gz`
`Error: Directory not empty @ dir_s_rmdir - /private/tmp/d20220127-58025-li20a1`

**Reply by gromgit** [reply]
> `Error: Directory not empty @ dir_s_rmdir - /private/tmp/d20220127-58025-li20a1`

That means Homebrew tried to clean up its temp directory after pouring the `openssl@1.1` bottle, but something else seems to have left something unexpected in that directory.

Are you running an anti-virus program or some other management solution? If so, try disabling it temporarily and re-run your `brew install`.

**Reply by name99-org** [reply]
I see the same behavior. Note
(a) this is MacOS 12.3.1. No anti-virus or similar such nonsense.
(b) the directory at fault is in /tmp. /tmp is, of course, owned by root
(c) I, as my user, not root, can create and delete folders in /tmp
(d) I can manually delete this problematic folder left by SSL openssl@1.1--1.1.1o.monterey.bottle.tar.gz

This looks like either a fault in this particular openssl installer or a fault in the design of dir_s_rmdir, whatever that is. 
The actual folder in /tmp left by openssl looks like it's essentially empty (a tree of empty folders) EXCEPT for the path openssl@1.1/1.1.1o/share/doc/openssl/html/man1 
which contains a whole bunch of html files. 
Same for man3, man5, man7

macOS does not use html man files, so my best guess would be something like there's a step that tries to move these html man files, fails, and doesn't delete them from this tmp location?

**Reply by tmillr** [reply]
> I have the same issue, but none of the above fix it:

same

```
==> `brew cleanup` has not been run in the last 30 days, running now...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Removing: ~/Library/Caches/Homebrew/pango--1.50.6... (773.8KB)
Removing: /usr/local/Cellar/sqlite/3.38.5... (11 files, 4.4MB)
Removing: ~/Library/Caches/Homebrew/sqlite--3.38.5... (2MB)
Error: Directory not empty @ dir_s_rmdir - /usr/local/lib/node_modules/@typescript-eslint/.DS_Store
```

**Comment by nana-marinbio** [comment]
I had the same issue trying to brew install qemu on macos:

`==> Installing qemu dependency: libffi
==> ./configure --prefix=/usr/local/Cellar/libffi/3.4.2 --libdir=/usr/local/Cellar/libffi/3.4.2/lib
==> make install
Error: Directory not empty @ dir_s_rmdir - /usr/local/opt/libffi`

It was the second trial. When I first tryied I had a python version error, so I upgraded python and tryied:
 `brew install --ignore-dependencies`  but I received a warning and the installation proceeded including all dependencies:
`Warning: `--ignore-dependencies` is an unsupported Homebrew developer flag!`

Read all the comments but still not sure how to solve! 
Any suggestions are welcome!

**Comment by lab0ryu** [comment]
Hi, I met the same problem, when I upgrade rust. 

`brew upgrade -vvv --debug rust`

output is: 
```text
==> Pouring rust--1.72.1.ventura.bottle.tar.gz
/usr/bin/env tar --extract --no-same-owner --file /Users/admin/Library/Caches/Homebrew/downloads/c4bfa52fc0f680e05b47c148405af572de928e47545086e475b269a423f5637c--rust--1.72.1.ventura.bottle.tar.gz --directory /private/tmp/d20230925-46127-16k6qmw
/usr/bin/env cp -pR /private/tmp/d20230925-46127-16k6qmw/.DS_Store /usr/local/Cellar/.DS_Store
/usr/bin/env cp -pR /private/tmp/d20230925-46127-16k6qmw/rust/. /usr/local/Cellar/rust
ln -s ../Cellar/rust/1.72.0_1/bin/cargo cargo
ln -s ../Cellar/rust/1.72.0_1/bin/cargo-clippy cargo-clippy
ln -s ../Cellar/rust/1.72.0_1/bin/clippy-driver clippy-driver
ln -s ../Cellar/rust/1.72.0_1/bin/rust-demangler rust-demangler
ln -s ../Cellar/rust/1.72.0_1/bin/rust-gdb rust-gdb
ln -s ../Cellar/rust/1.72.0_1/bin/rust-gdbgui rust-gdbgui
ln -s ../Cellar/rust/1.72.0_1/bin/rust-lldb rust-lldb
ln -s ../Cellar/rust/1.72.0_1/bin/rustc rustc
ln -s ../Cellar/rust/1.72.0_1/bin/rustdoc rustdoc
ln -s ../../Cellar/rust/1.72.0_1/share/doc/rust rust
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-add.1 cargo-add.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-bench.1 cargo-bench.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-build.1 cargo-build.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-check.1 cargo-check.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-clean.1 cargo-clean.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-doc.1 cargo-doc.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-fetch.1 cargo-fetch.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-fix.1 cargo-fix.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-generate-lockfile.1 cargo-generate-lockfile.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-help.1 cargo-help.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-init.1 cargo-init.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-install.1 cargo-install.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-locate-project.1 cargo-locate-project.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-login.1 cargo-login.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-logout.1 cargo-logout.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-metadata.1 cargo-metadata.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-new.1 cargo-new.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-owner.1 cargo-owner.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-package.1 cargo-package.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-pkgid.1 cargo-pkgid.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-publish.1 cargo-publish.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-remove.1 cargo-remove.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-report.1 cargo-report.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-run.1 cargo-run.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-rustc.1 cargo-rustc.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-rustdoc.1 cargo-rustdoc.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-search.1 cargo-search.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-test.1 cargo-test.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-tree.1 cargo-tree.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-uninstall.1 cargo-uninstall.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-update.1 cargo-update.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-vendor.1 cargo-vendor.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-verify-project.1 cargo-verify-project.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-version.1 cargo-version.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo-yank.1 cargo-yank.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/cargo.1 cargo.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/rustc.1 rustc.1
ln -s ../../../Cellar/rust/1.72.0_1/share/man/man1/rustdoc.1 rustdoc.1
ln -s ../../../Cellar/rust/1.72.0_1/share/zsh/site-functions/_cargo _cargo
ln -s ../Cellar/rust/1.72.0_1/lib/librustc_driver-d7af7c52e7c23764.dylib librustc_driver-d7af7c52e7c23764.dylib
ln -s ../Cellar/rust/1.72.0_1/lib/libstd-e87ac324b31d6dd5.dylib libstd-e87ac324b31d6dd5.dylib
ln -s ../Cellar/rust/1.72.0_1/lib/libtest-75b68e94c2a5ca56.dylib libtest-75b68e94c2a5ca56.dylib
ln -s ../Cellar/rust/1.72.0_1/lib/rustlib rustlib
Error: Directory not empty @ dir_s_rmdir - /private/tmp/d20230925-46127-16k6qmw
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1432:in `rmdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1432:in `block in remove_dir1'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1443:in `platform_support'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1431:in `remove_dir1'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1424:in `remove'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:760:in `block in remove_entry'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1481:in `ensure in postorder_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:1481:in `postorder_traverse'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/fileutils.rb:758:in `remove_entry'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/tmpdir.rb:101:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/unpack_strategy.rb:128:in `extract_nestedly'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/_methods.rb:270:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:110:in `stage'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:40:in `stage'
/usr/local/Homebrew/Library/Homebrew/software_spec.rb:384:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:1259:in `block in pour'
/usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:299:in `block in cd'
/usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:299:in `chdir'
/usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:299:in `cd'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/_methods.rb:270:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:1258:in `pour'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/_methods.rb:270:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:437:in `install'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/call_validation.rb:157:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/sorbet-runtime-0.5.10461/lib/types/private/methods/_methods.rb:270:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/upgrade.rb:218:in `install_formula'
/usr/local/Homebrew/Library/Homebrew/upgrade.rb:191:in `upgrade_formula'
/usr/local/Homebrew/Library/Homebrew/upgrade.rb:101:in `block in upgrade_formulae'
/usr/local/Homebrew/Library/Homebrew/upgrade.rb:100:in `each'
/usr/local/Homebrew/Library/Homebrew/upgrade.rb:100:in `upgrade_formulae'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:214:in `upgrade_outdated_formulae'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:136:in `upgrade'
/usr/local/Homebrew/Library/Homebrew/brew.rb:94:in `<main>'
```
